cornice/ext/swagger.py

```diff
@@ -0,0 +1,370 @@
+    1) Values are None if they weren't in guerystring
+    2) Values are lists only if corresponding Parameter's collectionFormat is 'multi', else provides last value
+
+    Multidict is a WebOb structure used by Pyramid:
+    https://webob.readthedocs.org/en/latest/#multidict
+
+    Matchdict is a Pyramid structure:
+    http://docs.pylonsproject.org/projects/pyramid/en/latest/narr/urldispatch.html#matchdict
+    """
+    qs = request.params  # querystring multiduct
+    match_dict = dict()
+    for param in parameters:
+        if param['in'] == 'query':
+            # Handle "multi" collectionTypes in a special way
+            if 'collectionType' in param and param['collectionType'] == 'multi':
+                match_dict[param['name']] = qs.getall(param['name'])
+            elif param['name'] in qs:
+                match_dict[param['name']] = qs[param['name']]
+            else:
+                match_dict[param['name']] = None
+    return match_dict
+
+def col2swag_type(col_type):
+    """ Converts a primitive type used by SQLAlchemy Columns, to Swagger types
+    """
+    col_type = col_type.lower()
+    if col_type in ['text', 'string']:
+        return {"type": "string"}
+    elif col_type in ['float', 'numeric', 'real']:
+        return {"type": "number"}
+    elif col_type in ['int', 'bigint', 'biginteger', 'smallinteger' ]:
+        return {"type": "integer"}
+    elif col_type in ['bool', 'boolean']:
+        return {"type": 'boolean'}
+    elif col_type[-2:] == '[]':
+        return {"type": "array", "items": col2swag_type(col_type[:-2])}
+    else:
+        return {"type": 'string'}
+
+def sqa2swag_model(models):
+    """ Interprets SQL Alchemy model and makes a Swagger Response Model
+    Code based on dnordberg's code at https://gist.github.com/dnordberg/5661696
+    """
+    # All DB columns WILL come back, so make them required
+    required = []
+    json_model = {}
+    json_properties = {}
+    title = []
+    if not isinstance(models, list):
+        models = [models]
+    for model in models:
+        # We may be using a metadata table without the __table__ attribute, so plan accordingly
+        if hasattr(model, '__table__'):
+            table = model.__table__
+        else:
+            table = model
+        properties = table.columns.items()
+        title.append(str(table))
+        for propkey, prop in properties:
+            try:
+                json_properties[propkey] = col2swag_type(str(prop.type))
+                required.append(propkey)
+            except:
+                pass
+    json_model["properties"] = json_properties
+    json_model['required'] = list(set(required))
+    json_model['title'] = str(title)
+    return json_model
+
+#TODO make multidict2matchdict for schema objects.
+
+##############################
+#  SWAGGER SPEC GENERATORS   #
+##############################
+def generate_swagger_spec(services, title, version, **kwargs):
```

Comment:
Oh wow, this is a big function with a lot of things inside. We should make it smaller grained :)